Lib/test/test_fractions.py

```diff
@@ -346,6 +346,26 @@ def testConversions(self):
 
         self.assertTypedEquals(0.1+0j, complex(F(1,10)))
 
+    def testBoolGuarateesBoolReturn(self):
+        # Ensure that __bool__ is used on numerator which guarantees a bool
+        # return.  See also bpo-39274.
+        class custom_value(int):
+            def __init__(self, value):
+                self.value = value
+
+            def __bool__(self):
+                self.bool_used = True
+                return bool(self.value)
+
+        r = F(1, 1)
+        r._numerator = custom_value(1)
```

Comment:
I would prefer to avoid accessing private attribute _numerator:

```
numerator = custom_value(1)
r = F(numerator, 1)
...
self.assertEqual(numerator.bool_used, True) 
```

(same below)

Reply:
I agree fixed, I just got lost a bit because I missed that `numerator.numerator` is used (which loses the original type), but since its only that it seems nicer. Of course if the instantiation logic changes, there is a chance of the test needing to be revised.